### PR TITLE
Monster ted

### PR DESCRIPTION
**Pull request by Ghostvv**
**Proposed changes**:
- refactor entities in ted
- part of https://github.com/RasaHQ/rasa/issues/6670

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by tabergma** [comment]
why not merge into `e2e` directly?

File: rasa/core/featurizers/single_state_featurizer.py [hunk 1]
```diff
@@ -273,6 +240,40 @@ def encode_state(
 
         return state_features
 
+    def encode_entity(
```
Comment (tabergma) [inline rasa/core/featurizers/single_state_featurizer.py:243]: we need to add a docstring

File: rasa/core/featurizers/tracker_featurizers.py [hunk 2]
```diff
@@ -91,6 +91,34 @@ def _convert_labels_to_ids(
...
+
+    @staticmethod
+    def _entity_data(event: UserUttered) -> Dict[Text, Any]:
+        if event.text:
+            return {TEXT: event.text, ENTITIES: event.entities}
+
+        # input is not textual, so add empty dict
+        return {}
+
+    def training_states_actions_and_entities(
```
Comment (tabergma) [inline rasa/core/featurizers/tracker_featurizers.py:115]: docstring

File: rasa/core/policies/ted_policy.py [hunk 3]
```diff
@@ -956,23 +1007,25 @@ def _encode_real_features_per_attribute(
...
             # boolean mask returns flat tensor
-            _sequence_lengths = tf.expand_dims(_sequence_lengths, axis=-1)
+            sequence_lengths = tf.expand_dims(sequence_lengths, axis=-1)
 
             mask_sequence_text = tf.squeeze(
-                self._compute_mask(_sequence_lengths), axis=1
+                self._compute_mask(sequence_lengths), axis=1
             )
-            sequence_lengths = _sequence_lengths + 1
+            sequence_lengths = sequence_lengths + 1
```
Comment (tabergma) [inline rasa/core/policies/ted_policy.py:1028]:
```suggestion
            sequence_lengths += 1
```

File: rasa/core/policies/ted_policy.py [hunk 4]
```diff
@@ -989,6 +1042,44 @@ def _encode_real_features_per_attribute(
...
+                    ).values
+                    last_dialogue_mask = tf.math.logical_not(
+                        tf.cast(
+                            tf.concat(
+                                [dialogue_indices, tf.zeros((1,), dtype=tf.int32)],
+                                axis=0,
+                            )[1:],
+                            dtype=tf.bool,
+                        )
+                    )
```
Comment (tabergma) [inline rasa/core/policies/ted_policy.py:1070]: Took me some time to follow the logic, but very smart 💯

File: rasa/core/policies/ted_policy.py [hunk 5]
```diff
@@ -989,6 +1042,44 @@ def _encode_real_features_per_attribute(
...
             if attribute == TEXT:
                 text_transformer_output = attribute_features
+                text_sequence_lengths = sequence_lengths
+
+                if self.use_only_last_dialogue_turn:
+                    # get the location of all last dialogue inputs
+                    dialogue_lengths = tf.cast(
+                        tf_batch_data[DIALOGUE][LENGTH][0], dtype=tf.int32
+                    )
+                    # TODO precompute dialogue_indices after creation of tf_batch_data
```
Comment (tabergma) [inline rasa/core/policies/ted_policy.py:1052]: should we quickly take care of this TODO?

File: rasa/core/policies/ted_policy.py [hunk 6]
```diff
@@ -1340,7 +1251,126 @@ def _process_batch_data(
...
+
+    def _batch_loss_entities(
+        self,
+        tf_batch_data: Dict[Text, Dict[Text, List[tf.Tensor]]],
+        dialogue_transformer_output: tf.Tensor,
+        text_transformer_output: tf.Tensor,
+        text_sequence_lengths: tf.Tensor,
+    ) -> tf.Tensor:
+
+        return tf.cond(
```
Comment (tabergma) [inline rasa/core/policies/ted_policy.py:1331]: please add a comment why we need this `tf.cond`

**Review by tabergma — COMMENTED** [review]
Thanks for taking this over and making it work! 🚀 

Left a couple of comments. What is the exact plan? Merge into my branch and then fix tests and clean up the remaining stuff or should we directly merge into `e2e`?

**Comment by Ghostvv** [comment]
> why not merge into e2e directly?
It is directly out of your branch, so either we close your PR, and merge this one into e2e, or merge into yours. It doesn't really matter.
I created this PR to your branch so that you could take a look at the differences, I introduced

**Comment by tabergma** [comment]
Ok, let's close my PR and merge this one directly into `e2e`.

File: rasa/core/policies/ted_policy.py (hunk 5 above)
Comment (Ghostvv) [reply rasa/core/policies/ted_policy.py:1053]: yes, should be done as part of this PR

File: rasa/core/policies/ted_policy.py (hunk 4 above)
Comment (Ghostvv) [reply rasa/core/policies/ted_policy.py:1071]: it's the zeros 😅, it's always a `0`😂

Comment (Ghostvv) [reply rasa/core/policies/ted_policy.py:1072]: which reminds me, I need to add a comment

**Comment by Ghostvv** [comment]
@tabergma I think, I fixed the relevant tests. There are a bunch of todos, but I think it is better to tackle them in separate PR, could you please give it a final review

**Comment by Ghostvv** [comment]
Do you know what this security scan is? that fails?

**Comment by tabergma** [comment]
Not sure about the failing security check. I guess we need to ask Engineering.

**Review by tabergma — APPROVED** [review]
Looks good 🚀  Left a few minor comments. 

I agree, let's tackle the TODOs in a separate PR. Also we should fix all the tests on `e2e` branch again.

File: examples/e2ebot/config.yml [hunk 7]
```diff
@@ -9,8 +9,8 @@ pipeline:
     analyzer: char_wb
     min_ngram: 1
     max_ngram: 4
-  - name: DIETClassifier
-    epochs: 200
+#  - name: DIETClassifier
+#    epochs: 200
```
[the remark this answers is not in the document]
Comment (Ghostvv) [reply examples/e2ebot/config.yml:13]: oh, it should be uncommented. The idea is that both is supported

File: rasa/core/policies/ted_policy.py [hunk 8]
```diff
@@ -757,16 +861,39 @@ def _prepare_encoding_layers(self, name: Text) -> None:
...
 
+    # ---GRAPH BUILDING HELPERS---
+
+    @staticmethod
+    def _compute_dialogue_indices(
+        tf_batch_data: Dict[Text, Dict[Text, List[tf.Tensor]]]
+    ) -> None:
+        dialogue_lengths = tf.cast(tf_batch_data[DIALOGUE][LENGTH][0], dtype=tf.int32)
+        # wrap in a list, because that's the structure of tf_batch_data
+        tf_batch_data[DIALOGUE][IDS] = [
```
[the remark this answers is not in the document]
Comment (Ghostvv) [reply rasa/core/policies/ted_policy.py:875]: what do you mean? to me `ids` and `indices` are the same thing

[the remark this answers is not in the document]
Comment (Ghostvv) [reply rasa/core/policies/ted_policy.py:875]: I guess it'll be better to create new constant for `indices`

File: examples/e2ebot/data/stories.yml [hunk 9]
```diff
@@ -10,7 +10,7 @@ stories:
 
 - story: sad path (text to text)
   steps:
-  - user: "Hello"
+  - user: "[Hello](bla)"
```
[the remark this answers is not in the document]
Comment (Ghostvv) [reply examples/e2ebot/data/stories.yml:13]: added to domain, also added to the list of e2e things

**Pull request merged**